### ISSUE TITLE
zpool remove <pool> <log device> returns 'pool already exists'

### ISSUE THREAD

**Issue by janlam7** [openzfs/zfs#4270]
I'm trying to remove a log device from my pool and already spent some time with @ilovezfs trying several things, like turning the log into a mirror and then removing it. (which 
@ilovezfs suggested trying it in FreeBSD, it gives the same error. But now, when I try it in ZoL it does NOT return the error anymore, but also does not remove the log device.

mediacenter genkernel # zpool remove pool4 sda3
cannot remove sda3: pool already exists

Let me know if you want to help debug this, or need information. Otherwise I'll probably be re-installing the pool somewhere this week, after I make sure my backups are fully up to date ;-)

Just noticed it's very similar to https://github.com/zfsonlinux/zfs/issues/1422

I'm using gentoo zfs-0.6.5.3-r3 with a patch for issue 4112.

**Comment by ryao** [comment]
Truly using the device guid. You can get it from the output of `zdb`.

**Comment by ilovezfs** [comment]
We did. Same error.

**Comment by janlam7** [comment]
I'm sorry to say that I needed to repartition the disk the log was on, and since I could not remove it I got rid of the pool and restored from a backup.

I did a fresh install with 0.6.5.3-r3, so I cannot reproduce it anymore.

I think I spent about 2 hours with @ilovezfs trying to get rid of the log device. It might be in the #zfsonlinux chatlog if you scroll up ;-)

**Comment by ilovezfs** [comment]
@janlam7 That's a shame. Pool destruction is never the outcome anyone wants to see.

**Comment by LuciusMartius** [comment]
I have had this problem for a while. I wanted to remove my SSD log device to use it elsewhere.
It always just prints
`cannot remove /dev/sdi: pool already exists`

What is this error message trying to tell me? Even if this behavior is somehow correct, the content of this error message is just plain nonsense.

I already tried the exporting the pool, physically removing the drive and reimporting it. The pool ran without problems, but the remove command still fails with the same error above.

Then I tried replacing it with another disk (rotational instead of SSD) which went fine but after resilvering finished, it again failes with the same message above.

Now I have just been using the pool without the log device plugged in for a few days and everything seems fine. I just have to import it with "-m" every time.

```
  pool: tank
 state: DEGRADED
status: One or more devices could not be used because the label is missing or
        invalid.  Sufficient replicas exist for the pool to continue
        functioning in a degraded state.
action: Replace the device using 'zpool replace'.
   see: http://zfsonlinux.org/msg/ZFS-8000-4J
  scan: resilvered 0 in 1h7m with 0 errors on Sat Jul  8 03:05:14 2017
config:

        NAME                   STATE     READ WRITE CKSUM
        tank                   DEGRADED     0     0     0
          raidz2-0             ONLINE       0     0     0
            sdc                ONLINE       0     0     0
            sdd                ONLINE       0     0     0
            sdf                ONLINE       0     0     0
            sde                ONLINE       0     0     0
            sda                ONLINE       0     0     0
            sdb                ONLINE       0     0     0
        logs
          2050186627143941104  UNAVAIL      0     0     0  was /dev/sdi1
```

errors: No known data errors

Would it be possible to add a "zpool remove -f <pool> <device>" flag for exactly these cases?
To enable the admin to just simply forcibly remove devices the pool is clearly running fine without?

**Comment by LuciusMartius** [comment]
Just for future reference: I repaired my pool by hacking the kernel module and disabling the checks that caused the remove to fail. For the chance that it is any help in debugging the issue, I'll document the changes I made below.

First I was able to offline the device by removing this check:
```diff --git a/module/zfs/zil.c b/module/zfs/zil.c
--- a/module/zfs/zil.c
+++ b/module/zfs/zil.c
@@ -1981,12 +1981,6 @@ zil_suspend(const char *osname, void **cookiep)
 	mutex_enter(&zilog->zl_lock);
 	zh = zilog->zl_header;
 
-	if (zh->zh_flags & ZIL_REPLAY_NEEDED) {		/* unplayed log */
-		mutex_exit(&zilog->zl_lock);
-		dmu_objset_rele(os, suspend_tag);
-		return (SET_ERROR(EBUSY));
-	}
-
 	/*
 	 * Don't put a long hold in the cases where we can avoid it.  This
 	 * is when there is no cookie so we are doing a suspend & resume
-- 
```
With this I was able to offline the device. But when I tried to remove it the module panicked.
So I also had to remove this check too:

```diff --git a/module/zfs/vdev.c b/module/zfs/vdev.c
--- a/module/zfs/vdev.c
+++ b/module/zfs/vdev.c
@@ -2233,7 +2233,6 @@ vdev_remove(vdev_t *vd, uint64_t txg)
 			 */
 			metaslab_group_histogram_remove(mg, msp);
 
-			VERIFY0(space_map_allocated(msp->ms_sm));
 			space_map_free(msp->ms_sm, tx);
 			space_map_close(msp->ms_sm);
 			msp->ms_sm = NULL;
-- 
```
After this I was able to remove the device. A subsequent scrub showed no errors to the pool so I'll call this a win.

Disclaimer: If anyone tries to do the same, make sure you have exactly the same problem. This will remove the log device despite data still (seeming to be) allocated and is **potentially very dangerous**. Also I suggest to import the pool from a live system, so there will be nothing potentially writing data to the log device when it's offlined. I booted to my initramfs busybox environment for this.

**Comment by mgmartin** [comment]
I just hit this on two separate pools.  Using latest spl/zfs  from Jan. 12 2018.

```
 pool: array1
 state: ONLINE
  scan: scrub repaired 0B in 0 days 11:29:16 with 0 errors on Wed Dec 20 04:33:40 2017
config:

	NAME                                           STATE     READ WRITE CKSUM
	array1                                         ONLINE       0     0     0
	  raidz1-0                                     ONLINE       0     0     0
	    wwn-0x50014ee05878a760                     ONLINE       0     0     0
	    wwn-0x50014ee0adcd6257                     ONLINE       0     0     0
	    ata-WDC_WD2002FAEX-007BA0_WD-WMAY04091025  ONLINE       0     0     0
	    wwn-0x50014ee05878a56e                     ONLINE       0     0     0
	    wwn-0x50014ee05878abf1                     ONLINE       0     0     0
	logs
	  sde1                                         ONLINE       0     0     0

errors: No known data errors

  pool: probox
 state: ONLINE
  scan: scrub canceled on Fri Nov 17 22:22:00 2017
config:

	NAME                        STATE     READ WRITE CKSUM
	probox                      ONLINE       0     0     0
	  raidz2-0                  ONLINE       0     0     0
	    wwn-0x50014ee057b465a4  ONLINE       0     0     0
	    wwn-0x50014ee2054b1e64  ONLINE       0     0     0
	    wwn-0x50014ee2b924e6e7  ONLINE       0     0     0
	    wwn-0x50014ee2b924497d  ONLINE       0     0     0
	    wwn-0x50014ee2b9241b1c  ONLINE       0     0     0
	    wwn-0x50014ee20e795e5c  ONLINE       0     0     0
	    wwn-0x50014ee2b9240e6d  ONLINE       0     0     0
	    wwn-0x50014ee263ce8355  ONLINE       0     0     0
	logs
	  sde2                      ONLINE       0     0     0

>zpool remove array1 sde1
cannot remove sde1: pool already exists
>zpool remove probox sde2
cannot remove sde2: pool already exists
```

I tried using the full /dev/ paths and also using guid.  Same result each time.

**Comment by TonyHoyleRps** [comment]
I've just hit this bug on an imported pool from another system and even the code changes above can't fix it - they remove the bogus error message but the log is not removed.

If it's impossible to remove logs then the error message should say so, not have people scraping all over the internet for clues to what 'pool already exists' means.

**Comment by kpande** [comment]
@tcaputi can you look at this one? seems related to encryption.

**Comment by tcaputi** [comment]
I think there are a couple things going on here. The bug was first reported on 0.6.5.3 which was long before encryption was merged (after 0.7.0 was tagged). Since the pool was destroyed a long time ago, I'm not sure if that problem has been solved or not. @behlendorf might remember though.

Separately, the encryption patch had a small issue where the error message telling people to mount their encrypted datasets was omitted. This was fixed in 2ffd89fc back in June 2018.
